Add a guide for writing and running tests

matusmarhefka opened the pull request:
We are missing a guide on how to write tests and where to put them which makes it hard for contributors to add tests for their PRs.

WIP progress:
* updated `docs/contribute/contribute.adoc`
* updated `README.md` to ling to `docs/contribute/contribute.adoc`
* updated `docs/contribute/versioning.adoc`

matusmarhefka commented:
@mpreisler I believe that we don't support `1.0.x` anymore, is that correct? If so, do you know the exact date when it went unsupported so I can update `docs/contribute/versioning.adoc`?

yuumasato commented on README.md:
```diff
@@ -91,6 +91,11 @@ make install
 
+## Contributing
+
+If you would like to contribute to the OpenSCAP, either fix some issue or add a new feature, please check our [contribution guide](docs/contribute/contribute.adoc).
```
> **you can** fix some issue or add a new feature

I suggest to be less strict and more open.

We could also invite people to join the mailing list here and contribute to discussions. 
Only mention of the list I could find was in section `Choose the right branch`

matusmarhefka commented:
@openscap-ci test this please

matusmarhefka commented:
@openscap-ci test this please

mpreisler commented:
We use the title for WIP marker, I have migrated this ticket to that and removed the wip label.

matusmarhefka commented:
I forgot to remove the wip label, sorry for confusion. This PR should be ready for a review.

mpreisler commented on README.md:
```diff
@@ -91,6 +91,11 @@ make install
 
+## Contributing
+
+If you would like to contribute to the OpenSCAP, either fix some issue or add a new feature, please check our [contribution guide](docs/contribute/contribute.adoc). In case you would like to discuss anything or ask questions you can either join our FreeNode IRC channel called **#openscap** or our [mailing list](https://www.redhat.com/mailman/listinfo/open-scap-list).
```
the `either.. or..` wording sounds weird here. I am not a native speaker though

cipherboy replied:
I'd suggest:

> We welcome all contributions to the OpenSCAP project. If you would like to contribute, either by fixing existing issues or adding new features, please check out our [contribution guide](docs/contribute/contribute.adoc) for help getting started. If you would like to discuss anything, ask questions, or if you need additional help getting started, you can either send a message to our FreeNode IRC channel, **#openscap**, or to our [mailing list](https://www.redhat.com/mailman/listinfo/open-scap-list).

matusmarhefka replied:
Thanks, I will use the suggested text.

mpreisler commented on docs/contribute/contribute.adoc:
```diff
@@ -4,126 +4,169 @@ This document is meant for new OpenSCAP contributors and it describes how to
 issue from the link:https://github.com/OpenSCAP/openscap/issues[tracker]. Let's
-say that we want to fix the isuues with number _455_ and the fix will go into
-the _maint-1.0_ branch.
+say that you want to fix the isuues with number _455_ and the fix will go into
```
isUUes

mpreisler commented on docs/contribute/contribute.adoc:
```diff
@@ -4,126 +4,169 @@ This document is meant for new OpenSCAP contributors and it describes how to
+`maint-1.2` branch you can create a new branch from it, which we will call
+_fix_455_, where you can work on the fix. Remember that the name of the new
+branch will appear in the commit message when your fix is merged so choose the
+name wisely:
```
lol, I totally fail at this and provide random or funny or outright rude branch names

mpreisler commented on docs/contribute/contribute.adoc:
```diff
@@ -4,126 +4,169 @@ This document is meant for new OpenSCAP contributors and it describes how to
 [source,bash]
 ----
-# checkout to the maint-1.0 branch
-$ git checkout maint-1.0
```
this command was actually very important here. without it you might branch off of an old feature branch.

either tell the user to do `git checkout -b fix_455 maint-1.2` or do the explicit `git checkout maint-1.2` before the branching.

mpreisler commented on docs/contribute/contribute.adoc:
```diff
@@ -4,126 +4,169 @@ This document is meant for new OpenSCAP contributors and it describes how to
+[source,bash]
+----
+# create a new local branch to track the remote master branch
+$ git checkout -b master remotes/origin/master
```
you will already have a local master tracking the origin master. this command is superfluous at best

matusmarhefka replied:
Are you sure about that? I documented this because when I do the clean clone of my forked repo I don't have it:
```
$ git clone https://github.com/matusmarhefka/openscap
$ cd openscap && git branch -a
* maint-1.2
  remotes/origin/HEAD -> origin/maint-1.2
  remotes/origin/maint-1.2
  remotes/origin/master
  remotes/origin/tests_doc
```

mpreisler commented on docs/contribute/contribute.adoc:
```diff
@@ -4,126 +4,169 @@ This document is meant for new OpenSCAP contributors and it describes how to
 
-NOTE: Since we're fixing an issue number 455 make sure that your commit message
-contain a
+NOTE: Since you're fixing the issue number 455 make sure that your commit
```
drop `the`. I am not sure why but I am pretty sure it's either `an issue`, `the issue that blablabla` or `issue number XYZ`.

matusmarhefka replied:
I don't think it can be `an issue` as in this case we are referring to the specific issue with number 455, but if you are sure, I will remove `the`.

mpreisler commented on docs/contribute/contribute.adoc:
```diff
@@ -4,126 +4,169 @@ This document is meant for new OpenSCAP contributors and it describes how to
 
-
 Let's say that you've fixed the issue, made a few commits to your local fix_455
 branch and you think it's ready to be review by other contributors. Before you
```
ready to be review

mpreisler commented on docs/contribute/contribute.adoc:
```diff
@@ -4,126 +4,169 @@ This document is meant for new OpenSCAP contributors and it describes how to
-if your changes will be applicable and won't be in a conflict with some work that
-  other contributors could publish while you were working on the fix.
+if your changes will be applicable and won't be in a conflict with some work
+that other contributors could publish while you were working on the fix.
```
could have published

mpreisler commented on docs/contribute/contribute.adoc:
```diff
@@ -4,126 +4,169 @@ This document is meant for new OpenSCAP contributors and it describes how to
-on top of them. First we need to make sure that our local maint-1.0 branch is
-up-to date.
+If some other contributor pushed some code to the `maint-1.2` branch while you
+were working on the fix then it's necessary to fetch those changes and rebase
```
it's not necessary, it might be necessary if there are conflicts but we don't require ff-only merges AFAIK

matusmarhefka replied:
True.

mpreisler commented on docs/contribute/contribute.adoc:
```diff
@@ -4,126 +4,169 @@ This document is meant for new OpenSCAP contributors and it describes how to
-# add remote name 'upstream' to refer to the upstream's repository
+# add remote name 'upstream' to refer to the upstream's repository (this needs
+# to be done only once - next time you rebase you can skip this step)
 $ git remote add upstream git@github.com:OpenSCAP/openscap.git
```
this is not the right time to be doing this at all. this should go to repo setup

matusmarhefka replied:
Moved.

mpreisler commented on docs/contribute/contribute.adoc:
```diff
@@ -4,126 +4,169 @@ This document is meant for new OpenSCAP contributors and it describes how to
+# fetch the branches and their respective commits from the upstream repository
+$ git fetch upstream
+# merge the changes from upstream/maint-1.2 into your local maint-1.2 branch
+$ git merge upstream/maint-1.2
```
this is a very verbose way of pulling.

just do: `git checkout maint-1.2; git pull upstream maint-1.2`

matusmarhefka replied:
Ok, I will change it to pull instead of fetch/merge, but basically it does the same thing.

mpreisler commented on docs/contribute/contribute.adoc:
```diff
@@ -4,126 +4,169 @@ This document is meant for new OpenSCAP contributors and it describes how to
+[source,bash]
+----
+# push your changes to your remote forked repository (also see the note below)
 $ git push origin fix_455
```
this won't work without --force if you rebased

mpreisler commented on docs/contribute/contribute.adoc:
```diff
@@ -135,40 +178,53 @@ After the review is done and one or more experienced developers is complaining
-the commit, where the change is needed, and use commit's `--ammed` option to
-change the commit. You can use following steps to do that.
+If you just need to fix something (for example a typo) you need to go back to
+the commit where the change is needed and use commit's `--ammed` option to
```
ammed

mpreisler commented on docs/contribute/contribute.adoc:
```diff
@@ -135,40 +178,53 @@ After the review is done and one or more experienced developers is complaining
+# detele the fix_455 branch locally
 $ git branch -d fix_455
+# optionally also delete the fix_455 branch from your remote forked repository
+$ git push origin --delete fix_455
```
or you can just click the github button under the PR

matusmarhefka replied:
I think this way is better as it will reflect in both, your cloned repo and also your remote. If you just click the button it will remove a branch from your forked repo on GH but the reference to it will remain in your cloned repo and you will need to `git remote prune origin` to sync it.

mpreisler commented on docs/contribute/testing.adoc:
```diff
@@ -0,0 +1,423 @@
+library, its utilities and the test directories corresponding to these parts.
+Therefore, use your best judgement when deciding where to put the test for your
+pull request and if you are not sure don't be affraid to ask in the pull
+request, someone will definitely help you with that.
```
so you say here that we are not gonna discuss where to put tests but 2 sections later you describe that :) this might be a leftover

matusmarhefka replied:
Not really, the point of this should be to tell that we don't have a strict rule where to put tests (+ in the current state the test placement is a mess anyway :D ), I will rewrite it to be more clear.

mpreisler commented on docs/contribute/testing.adoc:
```diff
@@ -0,0 +1,423 @@
+. Create `Makefile.am` inside the `tests/API/XCCDF/foo` directory with the
+   following contents:
+[[app-listing]]
+[source,bash]
```
is explaining autotools really in scope? there are better resources for this, perhaps link autotools docs?

mpreisler commented on docs/contribute/testing.adoc:
```diff
@@ -0,0 +1,423 @@
+----
+
+[start=4]
+. Create an entry script, for example `all.sh`, inside the `tests/API/XCCDF/foo`
```
We hate all.sh and are trying to move away from them. It sucks for multiple reasons, the main one is that if `all.sh` takes a long time (which some of the tests do, ~2 minutes) you can't parallelize it. If anything fails in it it's tough to navigate logs, etc etc etc...

Please provide a good example. We are aiming for a lot of small quick tests over all.sh that takes forever.

mpreisler commented on docs/contribute/testing.adoc:
```diff
@@ -0,0 +1,423 @@
+test_init test_api_xccdf_foo.log
+
+test_run "Description 1" $srcdir/my_test_script.sh my_data.xml
+test_run "Description 2" test_foo
```
again, we are trying to do just one test_run per file. and I will say that I don't like the testing harness we have, it's ridiculous and just makes things harder... we should rely on cmake/autotools to collect the right outputs, not write bash magic to fix it...

mpreisler commented on docs/contribute/testing.adoc:
```diff
@@ -0,0 +1,423 @@
+
+NOTE: The **entry script** is responsible for running all the tests from a
+specific directory and it is defined in the `TESTS` variable in the directory's
+`Makefile.am`.
```
no, just no :D

jan-cerny replied:
I also prefer small tests over large all.sh.

mpreisler commented on docs/contribute/testing.adoc:
```diff
@@ -0,0 +1,423 @@
+----
+
+
+==== Setting shell attributes for your test scripts
```
Suggestion: Change this to "Best practices when writing bash tests"

mpreisler commented on docs/contribute/testing.adoc:
```diff
@@ -0,0 +1,423 @@
+
+
+==== Setting shell attributes for your test scripts
+It is probably good to set `pipefail` option for all your test scripts so you
```
it's not probably good, it's a good idea. don't make it sound weaker by adding probably/maybe/etc...

jan-cerny commented on docs/contribute/testing.adoc:
```diff
@@ -0,0 +1,423 @@
+link:../../tests/API/XCCDF/unittests[tests/API/XCCDF/unittests] directory as
+all the general XCCDF tests for `oscap xccdf eval` are located there. We will
+add our new test into this subdirectory.
+
```
Do you think we could mention that our unit tests are not unit tests? If somebody decides where to put a new test it could be confusing for him if he thinks that unit tests are unit tests.

Also, it would be better to rename the tests that are not unit tests to something else than `unittests` in the repository, but that is out of scope of this PR.

matusmarhefka replied:
@jan-cerny I think it would be better to rename the folder rather than documenting it. I can do it in this PR, what do you say about changing it from `unittests` to `general`?

jan-cerny replied:
@matusmarhefka I think this will cause problems when merging maint-1.2 to master. I prefer do not do that in this PR. We can do that in a separate PR later. Such a huge renaming might need further discussion.

jan-cerny commented on docs/contribute/testing.adoc:
```diff
@@ -0,0 +1,423 @@
+----
+test_run "DESCRIPTION" $srcdir/TEST_SCRIPT_FILE|TEST_FUNCTION ARG [ARG...]
+----
+The `$srcdir` variable contains the path to the directory with the entry script.
```
I think we could describe the $srcdir and $builddir more in detail somewhere and describe which problems they have. From my experience, I often use them wrong which causes that `make distcheck` doesn't work. I would be nice to describe the difference of `make check ` and `make distcheck` as well.

matusmarhefka replied:
@jan-cerny As this is the gnu autotools specific thing it might be better to add a reference to its documentation, for example the following link:
https://www.gnu.org/software/automake/manual/html_node/Checking-the-Distribution.html

If you have some problem which is not clear from the documentation though, you can describe it here and I will try to add it into the testing guide.

jan-cerny replied:
It is a nice link, but my problem is that we always use it wrong, we do not run make distcheck, then we submit a PR, and then we wonder why Jenkins fails.

jan-cerny replied:
not always, but frequently

jan-cerny commented on docs/contribute/testing.adoc:
```diff
@@ -0,0 +1,423 @@
+
+The `test_single_rule.sh` script will contain the following tests to check
+the functionality of the single rule evaluation feature:
+
```
Idea: Does some clever book exist that explains how to create the test cases?

matusmarhefka replied:
This might be an overkill for testing guide, IMO most people would not read a book just to write a simple test for their code.

jan-cerny commented on docs/contribute/versioning.adoc:
```diff
@@ -1,57 +1,99 @@
 	* 1.2.x: supported
 	* 1.1.x: unsupported since 2015-10-16
-	* 1.0.x: supported, frozen since 2016-01-07
+	* 1.0.x: unsupported
```
I hope I can answer instead of @mpreisler. It was canceled on 2017-06-01. See https://www.redhat.com/archives/open-scap-list/2017-June/msg00000.html

matusmarhefka replied:
Thx, I will mention it there.

jan-cerny commented on docs/contribute/testing.adoc:
```diff
@@ -0,0 +1,423 @@
+=== Running your new test
+To run your new test you first need to compile the OpenSCAP library. See the
+*Compilation* section in the link:../../README.md[README.md] for more details.
+Also you don't need to run all the tests, only tests in a direcotory where you
```
directory

jan-cerny commented on docs/contribute/testing.adoc:
```diff
@@ -0,0 +1,423 @@
+= Writing and running tests for OpenSCAP
+
+This document should help you with writing and running tests for your pull
+requests. It is recommended to add a new test for every new functionality or
```
Maybe we could improve wording of this:

It is recommended to add a new test if a new functionality is added or if a code that is not covered by any test is updated.

jan-cerny commented on docs/contribute/testing.adoc:
```diff
@@ -0,0 +1,423 @@
+----
+result="./results.xml"
+my_rule_="xccdf_com.example.www_rule_test"
+assert_exists 1 "//rule-result[@idref=\"$my_rule\"]/result[text()=\"fail\"]"
```
Just a side question: Does it work with single quotes instead of escaping double quotes?

matusmarhefka replied:
Yes it does.

matusmarhefka commented on README.md:
```diff
@@ -91,6 +91,11 @@ make install
 
+## Contributing
+
+We welcome all contributions to the OpenSCAP project. If you would like to contribute, either by fixing existing issues or adding new features, please check out our [contribution guide](docs/contribute/contribute.adoc) for help getting started. If you would like to discuss anything, ask questions, or if you need additional help getting started, you can either send a message to our FreeNode IRC ch ...
```
This was suggested by @cipherboy and I believe it should be grammatically correct.

cipherboy replied:
I think the first "for help getting started" could be changed "to get started". Both are fine grammatically and "for help getting started" gets a bit repetitive. Thanks @jan-cerny

matusmarhefka commented:
I have rewritten the guide based on the reviews and also removed the `all.sh` paradigm. It should be also easier to merge it into the `master` after some changes.

matusmarhefka commented:
@openscap-ci test this please

mpreisler commented:
I am aware of additional feedback but I think this PR is a huge step forward and we shouldn't hold up the merging. More improvements can be done in future PRs.

Thx for this great work!
